src/compile/data/formatparse.ts

```diff
@@ -52,77 +53,85 @@ export class ParseNode extends DataFlowNode {
             return;
           }
-          parse[fieldDef.field] = 'number';
+          implicit[fieldDef.field] = 'number';
         } else if (accessPathDepth(fieldDef.field) > 1) {
           // For non-date/non-number (strings and booleans), derive a flattened field for a referenced nested field.
           // (Parsing numbers / dates already flattens numeric and temporal fields.)
-          parse[fieldDef.field] = 'flatten';
+          implicit[fieldDef.field] = 'flatten';
         }
       });
     }
 
-    // Custom parse should override inferred parse
-    const data = model.data;
-    if (data && data.format) {
-      const p = data.format.parse;
-      if (p === null) {
-        return null;  // disable parsing completely
-      } else if (p) {
-        keys(p).forEach(field => {
-          parse[field] = p[field];
-        });
+    return this.makeWithAncestors(parent, {}, implicit, ancestorParse);
+  }
+
+  /**
+   * Creates a parse node from two sets of explicit and implicit parse and updates ancestorParse.
+   */
+  public static makeWithAncestors(parent: DataFlowNode, explicit: Dict<string>, implicit: Dict<string>, ancestorParse: AncestorParse) {
+    // We should not parse what has already been parsed in a parent (explicitly or implicitly) or what has been derived (maked as "derived").
+    for (const field of keys(implicit)) {
+      const parsedAs = ancestorParse.getWithExplicit(field);
+      if (parsedAs.value !== undefined) {
+        // We always ignore derived fields even if they are implicitly defined because we expects users to create the right types.
+        if (parsedAs.explicit || parsedAs.value === implicit[field] || parsedAs.value === 'derived') {
+          delete implicit[field];
+        } else {
+          log.warn(log.message.differentParse(field, implicit[field], parsedAs.value));
+        }
       }
     }
 
-    // We should not parse what has already been parsed in a parent
-    const modelParse = model.component.data.ancestorParse;
-    keys(modelParse).forEach(field => {
-      if (parse[field] !== modelParse[field]) {
-        log.warn(log.message.differentParse(field, parse[field], modelParse[field]));
+    for (const field of keys(explicit)) {
+      const parsedAs = ancestorParse.get(field);
+      if (parsedAs !== undefined) {
+        if (parsedAs === explicit[field]) {
+          delete explicit[field];
       } else {
-        delete parse[field];
-      }
-    });
-
-    // filter out parse == null (do not parse the field)
-    for (const field of keys(parse)) {
-      if (parse[field] === null) {
-        delete parse[field];
+          log.warn(log.message.differentParse(field, explicit[field], parsedAs));
+        }
       }
     }
 
-    if (keys(parse).length === 0) {
+    const parse = new Split(explicit, implicit);
+
+    // add the format parse from this model so that children don't parse the same field again
+    ancestorParse.copyAll(parse);
```

Comment:
So you merge this into parent, but still create a local parse node? 

I'm pretty confused.

Reply:
ancestorParse is not a parse node, it's an object to track what has been parsed already